presto-orc/src/main/java/com/facebook/presto/orc/reader/SliceDictionarySelectiveReader.java

```diff
@@ -384,32 +388,41 @@ public Block getBlock(int[] positions, int positionCount)
             return new RunLengthEncodedBlock(outputType.createBlockBuilder(null, 1).appendNull().build(), positionCount);
         }
 
-        wrapDictionaryIfNecessary();
+        // compact values(ids) array, and calculate 1) the slice sizeInBytes if materialized, and 2) number of nulls
+        long blockSizeInBytes = 0;
+        int nullsCount = 0;  // the nulls count for selected positions
+        int i = 0;
+        int j = 0;
+        while (i < positionCount && j < outputPositionCount) {
+            if (positions[i] != outputPositions[j]) {
+                j++;
+                continue;
+            }
 
-        if (positionCount == outputPositionCount) {
-            DictionaryBlock block = new DictionaryBlock(positionCount, dictionary, values);
+            int id = this.values[j];
+            values[i] = id;
 
-            values = null;
-            return block;
+            blockSizeInBytes += dictionaryOffsetVector[id + 1] - dictionaryOffsetVector[id];
+            nullsCount += (id == currentDictionarySize - 1 ? 1 : 0);
+
+            i++;
+            j++;
         }
 
-        int[] valuesCopy = new int[positionCount];
+        // If all selected positions are null, just return RLE block.
+        if (nullsCount == outputPositionCount) {
```

Comment:
This doesn't seem right. I think it should be `nullCount == positionCount`.

Reply:
Thanks Masha. Yes it should be positionCount.